### PR TITLE
issue #27773 Partial resolution

### PR DESCRIPTION
**Pull request by ChristopherCotnoir**

### CODE REVIEW

**Comment by sindhupriyaav** [comment]
Performed UAT: Observed that Issue is fixed in 'Quantities on Hand' screen but the issue still exists in following screens
1. Item Sites screen
2. Item Availability Workbench

Also information related to other item sites can be viewed/edited in following screens
1. General Ledger Transactions
2. Planned Orders

**Comment by ChristopherCotnoir** [comment]
Itemsite and Item Availability Workbench screens are dealt with by xtuple/private-extensions#797. Gil Moskowitz says to  leave General Ledger Transactions as is.

**Comment by lokeshsh** [comment]
Performed UAT: Observed that Issue is fixed in the below screens:
1. Item Sites screen
2. Item Availability Workbench
3. 'Quantities on Hand' 
But on clicking 'Query' button in 'Planned Orders' screen, observed that 'Error Retrieving Information' dialog box is displayed with DB log error. 

DB Log Error:
Mon Jun 13 12:40:36 2016 ERROR:  invalid reference to FROM-clause entry for table "whsinfo"
LINE 29:         whsinfo.warehous_code AS warehous_code,
                 ^
HINT:  Perhaps you meant to reference the table alias "supplywhsinfo".
(42P01) QPSQL: Unable to create query
SELECT *,
        plorevenue - plocost AS profit,
        formatMoney(plorevenue - plocost) AS f_profit,
        CASE WHEN plocost > plorevenue THEN 'error'
        END AS plorevenue_xtforegroundrole,
        CASE WHEN (planord_firm) THEN 'emphasis'
        END AS qtforegroundrole,
        'qty' AS planord_qty_xtnumericrole,
        'curr' AS plocost_xtnumericrole,
        'curr' AS plorevenue_xtnumericrole,
        'curr' AS profit_xtnumericrole,
        0 AS plocost_xttotalrole,
        0 AS plorevenue_xttotalrole,
        0 AS profit_xttotalrole
 FROM (
 SELECT planord_id, planord_itemsite_id,
        planord.*,
        (select plancode_code from plancode where plancode_id = itemsite.itemsite_plancode_id) as plancode,
        formatPloNumber(planord_id) AS ordernum,
        CASE WHEN (planord_type='P') THEN 'P/O'
             WHEN (planord_type='W') THEN 'W/O'
             WHEN (planord_type='T') THEN 'T/O'
             ELSE '?'
        END AS ordtype,
        formatDate(planord_duedate) AS f_duedate,
        formatDate(planord_startdate) AS f_startdate,
        formatQty(planord_qty) AS f_qty,
        formatboolYN(planord_firm) AS f_firm,
        whsinfo.warehous_code AS warehous_code,
        supplywhsinfo.warehous_code AS supply_warehous_code,
        item_number,
        (item_descrip1 || ' ' || item_descrip2) AS item_descrip,
        firstline(planord_comments) AS comments,
        uom_name,
        (classcode_code || '-' || classcode_descrip) AS f_classcode,
 
        0
  AS plocost,
 
         0
  AS plorevenue,
 
        formatMoney(0)
  AS f_plocost,
 
         formatMoney(0)
  AS f_plorevenue
 FROM planord JOIN itemsite ON (planord_itemsite_id=itemsite.itemsite_id)
              JOIN site() ON (itemsite.itemsite_warehous_id=warehous_id)
              JOIN item ON (itemsite.itemsite_item_id=item_id)
              JOIN uom ON (item_inv_uom_id=uom_id)
              LEFT OUTER JOIN classcode ON (classcode_id=item_classcode_id)
              LEFT OUTER JOIN itemsite supplyitemsite ON (planord_supply_itemsite_id=supplyitemsite.itemsite_id)
              LEFT OUTER JOIN whsinfo supplywhsinfo ON (supplyitemsite.itemsite_warehous_id=supplywhsinfo.warehous_id)
 
 WHERE (TRUE
  ) ) AS data
 ORDER BY planord_duedate, item_number;
Mon Jun 13 12:40:38 2016 Debug: Looking for a script  "QObject"
Mon Jun 13 12:40:38 2016 Debug: Looking for a script  "QWidget"
Mon Jun 13 12:40:38 2016 Debug: Looking for a script  "XWidget"
Mon Jun 13 12:40:38 2016 Debug: Looking for a script  "errorLog"
####Please refer the attached 'planned order.png' screenshot ####
planned order
'

**Comment by rkannadevara** [comment]
Performed UAT: Observed that issue still exists. On clicking 'Query' button in 'Planned Orders' screen displays 'Error Retrieving Information' error dialog.

Tue Jun 14 12:31:51 2016 ERROR:  column reference "warehous_code" is ambiguous
LINE 29:         warehous_code AS warehous_code,
                 ^
(42702) QPSQL: Unable to create query
SELECT *,
        plorevenue - plocost AS profit,
        formatMoney(plorevenue - plocost) AS f_profit,
        CASE WHEN plocost > plorevenue THEN 'error'
        END AS plorevenue_xtforegroundrole,
        CASE WHEN (planord_firm) THEN 'emphasis'
        END AS qtforegroundrole,
        'qty' AS planord_qty_xtnumericrole,
        'curr' AS plocost_xtnumericrole,
        'curr' AS plorevenue_xtnumericrole,
        'curr' AS profit_xtnumericrole,
        0 AS plocost_xttotalrole,
        0 AS plorevenue_xttotalrole,
        0 AS profit_xttotalrole
 FROM (
 SELECT planord_id, planord_itemsite_id,
        planord.*,
        (select plancode_code from plancode where plancode_id = itemsite.itemsite_plancode_id) as plancode,
        formatPloNumber(planord_id) AS ordernum,
        CASE WHEN (planord_type='P') THEN 'P/O'
             WHEN (planord_type='W') THEN 'W/O'
             WHEN (planord_type='T') THEN 'T/O'
             ELSE '?'
        END AS ordtype,
        formatDate(planord_duedate) AS f_duedate,
        formatDate(planord_startdate) AS f_startdate,
        formatQty(planord_qty) AS f_qty,
        formatboolYN(planord_firm) AS f_firm,
        warehous_code AS warehous_code,
        supplywhsinfo.warehous_code AS supply_warehous_code,
        item_number,
        (item_descrip1 || ' ' || item_descrip2) AS item_descrip,
        firstline(planord_comments) AS comments,
        uom_name,
        (classcode_code || '-' || classcode_descrip) AS f_classcode,
 
        0
  AS plocost,
 
         0
  AS plorevenue,
 
        formatMoney(0)
  AS f_plocost,
 
         formatMoney(0)
  AS f_plorevenue
 FROM planord JOIN itemsite ON (planord_itemsite_id=itemsite.itemsite_id)
              JOIN site() ON (itemsite.itemsite_warehous_id=warehous_id)
              JOIN item ON (itemsite.itemsite_item_id=item_id)
              JOIN uom ON (item_inv_uom_id=uom_id)
              LEFT OUTER JOIN classcode ON (classcode_id=item_classcode_id)
              LEFT OUTER JOIN itemsite supplyitemsite ON (planord_supply_itemsite_id=supplyitemsite.itemsite_id)
              LEFT OUTER JOIN whsinfo supplywhsinfo ON (supplyitemsite.itemsite_warehous_id=supplywhsinfo.warehous_id)
 
 WHERE (TRUE
 
 
 
 
 
 
 
 
  ) ) AS data
 ORDER BY planord_duedate, item_number;
Tue Jun 14 12:34:05 2016 ERROR:  column reference "warehous_code" is ambiguous
LINE 29:         warehous_code AS warehous_code,
                 ^
(42702) QPSQL: Unable to create query
SELECT *,
        plorevenue - plocost AS profit,
        formatMoney(plorevenue - plocost) AS f_profit,
        CASE WHEN plocost > plorevenue THEN 'error'
        END AS plorevenue_xtforegroundrole,
        CASE WHEN (planord_firm) THEN 'emphasis'
        END AS qtforegroundrole,
        'qty' AS planord_qty_xtnumericrole,
        'curr' AS plocost_xtnumericrole,
        'curr' AS plorevenue_xtnumericrole,
        'curr' AS profit_xtnumericrole,
        0 AS plocost_xttotalrole,
        0 AS plorevenue_xttotalrole,
        0 AS profit_xttotalrole
 FROM (
 SELECT planord_id, planord_itemsite_id,
        planord.*,
        (select plancode_code from plancode where plancode_id = itemsite.itemsite_plancode_id) as plancode,
        formatPloNumber(planord_id) AS ordernum,
        CASE WHEN (planord_type='P') THEN 'P/O'
             WHEN (planord_type='W') THEN 'W/O'
             WHEN (planord_type='T') THEN 'T/O'
             ELSE '?'
        END AS ordtype,
        formatDate(planord_duedate) AS f_duedate,
        formatDate(planord_startdate) AS f_startdate,
        formatQty(planord_qty) AS f_qty,
        formatboolYN(planord_firm) AS f_firm,
        warehous_code AS warehous_code,
        supplywhsinfo.warehous_code AS supply_warehous_code,
        item_number,
        (item_descrip1 || ' ' || item_descrip2) AS item_descrip,
        firstline(planord_comments) AS comments,
        uom_name,
        (classcode_code || '-' || classcode_descrip) AS f_classcode,
 
        0
  AS plocost,
 
         0
  AS plorevenue,
 
        formatMoney(0)
  AS f_plocost,
 
         formatMoney(0)
  AS f_plorevenue
 FROM planord JOIN itemsite ON (planord_itemsite_id=itemsite.itemsite_id)
              JOIN site() ON (itemsite.itemsite_warehous_id=warehous_id)
              JOIN item ON (itemsite.itemsite_item_id=item_id)
              JOIN uom ON (item_inv_uom_id=uom_id)
              LEFT OUTER JOIN classcode ON (classcode_id=item_classcode_id)
              LEFT OUTER JOIN itemsite supplyitemsite ON (planord_supply_itemsite_id=supplyitemsite.itemsite_id)
              LEFT OUTER JOIN whsinfo supplywhsinfo ON (supplyitemsite.itemsite_warehous_id=supplywhsinfo.warehous_id)
 
 WHERE (TRUE
 
 
 
 
 
 
 
 
  ) ) AS data
 ORDER BY planord_duedate, item_number;
Tue Jun 14 12:34:17 2016 ERROR:  column reference "warehous_code" is ambiguous
LINE 29:         warehous_code AS warehous_code,
                 ^
(42702) QPSQL: Unable to create query
SELECT *,
        plorevenue - plocost AS profit,
        formatMoney(plorevenue - plocost) AS f_profit,
        CASE WHEN plocost > plorevenue THEN 'error'
        END AS plorevenue_xtforegroundrole,
        CASE WHEN (planord_firm) THEN 'emphasis'
        END AS qtforegroundrole,
        'qty' AS planord_qty_xtnumericrole,
        'curr' AS plocost_xtnumericrole,
        'curr' AS plorevenue_xtnumericrole,
        'curr' AS profit_xtnumericrole,
        0 AS plocost_xttotalrole,
        0 AS plorevenue_xttotalrole,
        0 AS profit_xttotalrole
 FROM (
 SELECT planord_id, planord_itemsite_id,
        planord.*,
        (select plancode_code from plancode where plancode_id = itemsite.itemsite_plancode_id) as plancode,
        formatPloNumber(planord_id) AS ordernum,
        CASE WHEN (planord_type='P') THEN 'P/O'
             WHEN (planord_type='W') THEN 'W/O'
             WHEN (planord_type='T') THEN 'T/O'
             ELSE '?'
        END AS ordtype,
        formatDate(planord_duedate) AS f_duedate,
        formatDate(planord_startdate) AS f_startdate,
        formatQty(planord_qty) AS f_qty,
        formatboolYN(planord_firm) AS f_firm,
        warehous_code AS warehous_code,
        supplywhsinfo.warehous_code AS supply_warehous_code,
        item_number,
        (item_descrip1 || ' ' || item_descrip2) AS item_descrip,
        firstline(planord_comments) AS comments,
        uom_name,
        (classcode_code || '-' || classcode_descrip) AS f_classcode,
 
        0
  AS plocost,
 
         0
  AS plorevenue,
 
        formatMoney(0)
  AS f_plocost,
 
         formatMoney(0)
  AS f_plorevenue
 FROM planord JOIN itemsite ON (planord_itemsite_id=itemsite.itemsite_id)
              JOIN site() ON (itemsite.itemsite_warehous_id=warehous_id)
              JOIN item ON (itemsite.itemsite_item_id=item_id)
              JOIN uom ON (item_inv_uom_id=uom_id)
              LEFT OUTER JOIN classcode ON (classcode_id=item_classcode_id)
              LEFT OUTER JOIN itemsite supplyitemsite ON (planord_supply_itemsite_id=supplyitemsite.itemsite_id)
              LEFT OUTER JOIN whsinfo supplywhsinfo ON (supplyitemsite.itemsite_warehous_id=supplywhsinfo.warehous_id)
 
 WHERE (TRUE
 
 
 
 
 
 
 
 
  ) ) AS data
 ORDER BY planord_duedate, item_number;
plannedorder

File: foundation-database/public/tables/metasql/opensalesorders-detail.mql [hunk 1]
```diff
@@ -110,6 +101,13 @@ WHERE((true)
             WHEN <? value("holdtype") ?> = 3 THEN cohead_holdtype = 'P'
             WHEN <? value("holdtype") ?> = 4 THEN cohead_holdtype = 'R' END)
 <? endif ?>
+  AND   ((SELECT COUNT(*)
+         FROM coitem JOIN itemsite ON (itemsite_id=coitem_itemsite_id)
+                     JOIN site() ON (warehous_id=itemsite_warehous_id)
+        WHERE (coitem_cohead_id=cohead_id))=
+         (SELECT COUNT(*)
+         FROM coitem
+        WHERE (coitem_cohead_id=cohead_id))) )
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/opensalesorders-detail.mql]: This subselect ignores the case where the coitem does not have an itemsite. The following should do the same job and handle the extra case, but might be less expensive to run:
```sql
and not exists(select 1
                 from coitem
                 join itemsite on coitem_itemsite_id = itemsite_id
                 left outer join site() on itemsite_warehous_id = warehous_id
                where coitem_itemsite_id is not null
                  and warehous_id is null)
```

File: foundation-database/public/tables/metasql/purchaseOrders-detail.mql [hunk 2]
```diff
@@ -57,7 +57,7 @@ SELECT pohead_id, pohead_number,
 <? if exists("byReceiptDate") ?>
        recv,
 <? endif ?>
-       pohead LEFT OUTER JOIN whsinfo ON (pohead_warehous_id=warehous_id)
+       pohead JOIN site() ON (pohead_warehous_id=warehous_id)
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/purchaseOrders-detail.mql]: This doesn't show purchase orders that have no associated site.

File: foundation-database/public/tables/metasql/quotes-detail.mql [hunk 3]
```diff
@@ -29,7 +29,7 @@ SELECT quhead.*, firstline(quhead_ordercomments) AS notes, quhead_id AS id,
   CASE WHEN (quhead_expire < current_date) THEN
     'error'
    END AS quhead_expire_qtforegroundrole
-FROM quhead
+FROM quhead JOIN site() ON (quhead_warehous_id=warehous_id)
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/quotes-detail.mql]: This doesn't show quotes that have no associated site.

File: foundation-database/public/tables/metasql/salesOrders-detail.mql [hunk 4]
```diff
@@ -31,7 +31,7 @@ SELECT cohead_id, cust_id,
                                 AND (coitem_cohead_id=cohead_id))) > 0) ) THEN <? value("open") ?>
          ELSE <? value("partial") ?>
        END AS order_status
-FROM cohead 
+FROM cohead JOIN site() ON (cohead_warehous_id=warehous_id)
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/salesOrders-detail.mql]: This doesn't show sales orders that have no associated site.

File: foundation-database/public/tables/metasql/salesOrders-detail.mql [hunk 5]
```diff
@@ -75,7 +75,10 @@ WHERE ( (1=1)
   AND   ((SELECT COUNT(*)
          FROM coitem JOIN itemsite ON (itemsite_id=coitem_itemsite_id)
                      JOIN site() ON (warehous_id=itemsite_warehous_id)
-        WHERE (coitem_cohead_id=cohead_id)) > 0) )
+        WHERE (coitem_cohead_id=cohead_id))=
+         (SELECT COUNT(*)
+         FROM coitem
+        WHERE (coitem_cohead_id=cohead_id))) )
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/salesOrders-detail.mql]: See https://github.com/xtuple/xtuple/pull/2722/files#r66967256

File: foundation-database/public/tables/metasql/salesOrders-detail.mql (hunk 4 above)
Comment (ChristopherCotnoir) [inline foundation-database/public/tables/metasql/salesOrders-detail.mql]: Is that possible? The Sales Order screen controls site with a non-nullable dropdown, doesn't it?

Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/salesOrders-detail.mql]: The cohead table allows it and some orders might created through another interface, such as CSVImp or the REST API.

**Comment by bendiy** [comment]
@gilmoskowitz You mentioned a possible performance issue with this change. Let's set the `site()` function as `STABLE`. It's currently not set, so it defaults to `VOLATILE`, which doesn't cache the result. 
https://github.com/xtuple/xtuple/blob/4_10_x/foundation-database/public/functions/site.sql#L94-L96

https://www.postgresql.org/docs/9.3/static/sql-createfunction.html
> STABLE indicates that the function cannot modify the database, and that within a single table scan it will consistently return the same result for the same argument values, but that its result could change across SQL statements. This is the appropriate selection for functions whose results depend on database lookups, parameter variables (such as the current time zone), etc. (It is inappropriate for AFTER triggers that wish to query rows modified by the current command.) Also note that the current_timestamp family of functions qualify as stable, since their values do not change within a transaction.

File: foundation-database/public/tables/metasql/quotes-detail.mql [hunk 6]
```diff
@@ -58,7 +58,8 @@ FROM quhead JOIN site() ON (quhead_warehous_id=warehous_id)
   LEFT OUTER JOIN char char_alias<? literal("char_id_date_list") ?> 
              ON (charass_alias<? literal("char_id_date_list") ?>.charass_char_id=char_alias<? literal("char_id_date_list") ?>.char_id)
 <? endforeach ?>
-WHERE (checkQuoteSitePrivs(quhead_id, <? value("warehous_id") ?>))
+WHERE (quhead_warehous_id=warehous_id OR quhead_warehous_id IS NULL)
+      (checkQuoteSitePrivs(quhead_id, <? value("warehous_id") ?>))
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/quotes-detail.mql]: missing `AND`

**Comment by lokeshsh** [comment]
Performed UAT: Issue is fixed

**Pull request merged**